packages/stats/tests/test_gbstats.py

```diff
@@ -832,70 +755,99 @@ def test_format_results_denominator(self):
+            self.variation_results["main_sum_squares"]
+            - self.variation_results["users"] * self.variation_results["cr"] ** 2
+        ) / (self.variation_results["users"] - 1)
+        self.variation_results["lower"] = round(
+            self.variation_results["cr"]
+            - np.sqrt(self.variation_results["v"] / self.variation_results["users"])
+            * 1.959964,
+            4,
+        )
+        self.variation_results["upper"] = round(
+            self.variation_results["cr"]
+            + np.sqrt(self.variation_results["v"] / self.variation_results["users"])
+            * 1.959964,
+            4,
+        )
+
+    def test_update_message(self):
+        self.assertEqual(self.r.updateMessage, self.update_messages[0])
+
+    def test_single_variation_results(self):
+        if self.r.singleVariationResults:
+            cr_rounded = [
+                round(single_variation_result.cr, 4)
+                if single_variation_result.cr
+                else 0
+                for single_variation_result in self.r.singleVariationResults
+            ]
+            lower_rounded = [
+                round(single_variation_result.ci[0], 4)
+                if single_variation_result.ci
+                else 0
+                for single_variation_result in self.r.singleVariationResults
+            ]
+            upper_rounded = [
+                round(single_variation_result.ci[1], 4)
+                if single_variation_result.ci
+                else 0
+                for single_variation_result in self.r.singleVariationResults
+            ]
+            self.assertEqual(cr_rounded, list(self.variation_results["cr"]))
+            self.assertEqual(lower_rounded, list(self.variation_results["lower"]))
+            self.assertEqual(upper_rounded, list(self.variation_results["upper"]))
+        else:
+            assert 1 > 2, "singleVariationResults do not exist"
+
+    def test_variation_weights(self):
+        rng = np.random.default_rng(seed=self.bandit_analysis.bandit_weights_seed)
+        m = np.array(self.variation_results["cr"])
+        v = np.array(self.variation_results["v"] / self.variation_results["users"])
+        y = rng.multivariate_normal(
+            mean=m,
+            cov=np.diag(v),
+            size=int(1e4),
+        )
+        best_probs = np.empty((self.num_variations,))
+        best_rows = np.max(y, axis=1)
+        for variation in range(self.num_variations):
+            best_probs[variation] = np.mean(y[:, variation] == best_rows)
+        best_probs[best_probs < 0.01] = 0.01
+        best_probs /= sum(best_probs)
+        best_probs = np.round(np.array(best_probs, dtype=float), decimals=4)
+        if self.r.updatedWeights:
+            result_rounded = np.array([round(num, 4) for num in self.r.updatedWeights])
+            self.assertTrue(np.array_equal(result_rounded, best_probs))
         else:
-            assert 1 > 2, "wrong class"
+            self.assertEqual(1, 2)
```

Comment:
Is the point to throw an error if self.r.updatedWeights is false? Let's do that explicitly instead?

Reply:
good point, and done